syz-manager/snapshot.go

```diff
@@ -0,0 +1,200 @@
+// Copyright 2024 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package main
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"time"
+	"unsafe"
+
+	"github.com/google/flatbuffers/go"
+	"github.com/google/syzkaller/pkg/flatrpc"
+	"github.com/google/syzkaller/pkg/fuzzer/queue"
+	"github.com/google/syzkaller/pkg/log"
+	"github.com/google/syzkaller/vm"
+)
+
+func (mgr *Manager) snapshotLoop() {
+	queue.StatNumFuzzing.Add(mgr.vmPool.Count())
+
+	reqc := make(chan *queue.Request, 2*mgr.vmPool.Count())
+	go func() {
+		for {
+			reqc <- mgr.source.Next()
+		}
+	}()
+
+	errc := make(chan error)
+	for index := 0; index < mgr.vmPool.Count(); index++ {
+		index := index
+		go func() {
+			errc <- mgr.snapshotVM(index, reqc)
+		}()
+	}
+	log.Fatal(<-errc)
+}
+
+func (mgr *Manager) snapshotVM(index int, reqc <-chan *queue.Request) error {
+	inst, err := mgr.vmPool.Create(index)
+	if err != nil {
+		return err
+	}
+	executor, err := inst.Copy(mgr.cfg.ExecutorBin)
+	if err != nil {
+		return err
+	}
+	// All network connections (including ssh) will break once we start restoring snapshots.
+	// So we start a background process and log to /dev/kmsg.
+	cmd := fmt.Sprintf("nohup %v exec snapshot 1>/dev/null 2>/dev/kmsg </dev/null &", executor)
+	if _, _, err := inst.Run(time.Hour, mgr.reporter, cmd); err != nil {
+		return err
+	}
+
+	for first := true; ; first = false {
+		var req *queue.Request
+		select {
+		case req = <-reqc:
+		default:
+			start := time.Now()
+			req = <-reqc
+			queue.StatNoExecRequests.Add(1)
+			queue.StatNoExecDuration.Add(int(time.Since(start)))
+		}
+
+		if first {
+			if err := mgr.snapshotSetup(inst, req.ExecOpts.EnvFlags); err != nil {
+				return err
+			}
+		}
+
+		queue.StatExecs.Add(1)
+		res, output, err := mgr.snapshotRun(inst, req)
+		if err != nil {
+			return err
+		}
+
+		if mgr.reporter.ContainsCrash(output) {
+			res.Status = queue.Crashed
+			rep := mgr.reporter.Parse(output)
+			buf := new(bytes.Buffer)
+			fmt.Fprintf(buf, "program:\n%s\n", req.Prog.Serialize())
+			buf.Write(rep.Output)
+			rep.Output = buf.Bytes()
+			mgr.mu.Lock()
+			mgr.saveCrash(&Crash{Report: rep})
+			mgr.mu.Unlock()
+		}
+
+		req.Done(res)
+	}
+}
+
+func (mgr *Manager) snapshotSetup(inst *vm.Instance, env flatrpc.ExecEnv) error {
+	type handshakeReq struct {
```

Comment:
Did you consider the alternatives to defining and unsafe-casting a struct here? If we defined a `flatrpc` structure and packed/unpacked it with the standard flatrpc functionality, would it be much more code?

Reply:
I've considered, but it's a bit messy until refactor all executor to use flatrpc for internal communication (e.g. don't use these handshake_req/execute_req).

But done now.